nexus/src/app/background/instance_watcher.rs

```diff
@@ -0,0 +1,635 @@
+                InstanceUpdated {
+                    instance_updated: true,
+                    vmm_updated: false,
+                } => self.instance_updated.datum += 1,
+                InstanceUpdated {
+                    instance_updated: false,
+                    vmm_updated: true,
+                } => self.vmm_updated.datum += 1,
+                InstanceUpdated {
+                    instance_updated: false,
+                    vmm_updated: false,
+                } => self.no_update.datum += 1,
+            }
+            self.touched = true;
+        }
+
+        pub(super) fn check_failure(&mut self, reason: CheckFailure) {
+            self.check_failures
+                .entry(reason)
+                .or_insert_with(|| InstanceCheckFailures {
+                    reason: reason.to_string(),
+                    datum: Cumulative::default(),
+                })
+                .datum += 1;
+            self.touched = true;
+        }
+
+        pub(super) fn check_error(&mut self, reason: CheckError) {
+            self.update_failures
+                .entry(reason)
+                .or_insert_with(|| InstanceCheckErrors {
+                    reason: reason.to_string(),
+                    datum: Cumulative::default(),
+                })
+                .datum += 1;
+            self.touched = true;
+        }
+
+        fn len(&self) -> usize {
+            4 + self.check_failures.len() + self.update_failures.len()
+        }
+
+        fn sample_into(
+            &self,
+            target: &InstanceTarget,
+            dest: &mut Vec<Sample>,
+        ) -> Result<(), MetricsError> {
+            dest.push(Sample::new(target, &self.no_update)?);
+            dest.push(Sample::new(target, &self.instance_updated)?);
+            dest.push(Sample::new(target, &self.vmm_updated)?);
+            dest.push(Sample::new(target, &self.both_updated)?);
+            for metric in self.check_failures.values() {
+                dest.push(Sample::new(target, metric)?);
+            }
+            for metric in self.update_failures.values() {
+                dest.push(Sample::new(target, metric)?);
+            }
+            Ok(())
+        }
+    }
+
+    /// The number of successful checks for a single instance and sled agent
+    /// pair.
+    #[derive(Clone, Debug, Metric)]
+    struct InstanceChecks {
+        /// `true` if the instance state changed as a result of this check.
+        instance_updated: bool,
+        /// `true` if the VMM state changed as a result of this check.
+        vmm_updated: bool,
+        /// The number of successful checks for this instance and sled agent.
+        datum: Cumulative<u64>,
+    }
+
+    /// The number of failed checks for an instance and sled agent pair.
+    #[derive(Clone, Debug, Metric)]
+    struct InstanceCheckFailures {
+        /// The reason why the check failed.
+        ///
+        /// # Note
+        /// This must always be generated from a `CheckFailure` enum.
```

Comment:
(Here and below) I see this *happens* to come from a `.to_string()` call on `CheckFailure`, but why **must** it be this way? I wanna make sure I grok this, I'm not sure where that requirement is coming from?